library/std/src/os/unix/net/ancillary.rs

```diff
@@ -429,6 +441,34 @@ impl<'a> SocketAncillary<'a> {
         self.buffer.len()
     }
 
+    /// Returns the raw ancillary data as byte slice.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+    pub fn data(&self) -> &[u8] {
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buffer[..self.length]) }
+    }
+
+    /// Returns the entire buffer, including unused capacity.
+    ///
+    /// Use [`data()`](Self::data) if you are only interested in the used portion of the buffer.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+    pub fn buffer(&self) -> &[MaybeUninit<u8>] {
+        self.buffer
+    }
+
+    /// Returns the entire buffer as mutable slice, including unused capacity.
+    ///
+    /// You should normally call [`set_len()`](Self::set_len)
+    /// after changing the contents of the buffer.
+    ///
+    /// # Safety
+    /// All data written to the buffer must be valid ancillary data for the target platform,
+    /// and you must call [`set_len()`](Self::set_len) after changing
+    /// the buffer contents to update the internal bookkeeping.
```

Comment:
`cmsg(3)` says

>          When initializing a buffer that will contain a series of
>          cmsghdr structures (e.g., to be sent with sendmsg(2)), that
>          buffer should first be zero-initialized to ensure the correct
>          operation of CMSG_NXTHDR().

I think the equivalent should be mentioned in the doc. writing structs to the buffer may be insufficient since they might contain padding that may or may not get copied.

Ultimately users would have to reimplement `add_to_ancillary_data` to get this right.

Reply:
Hmm, for the use case where the kernel reads into the buffer this wouldn't matter.

I think this only becomes a problem when you fill the buffer with data manually, and then use the iterator API. That's a weird thing to do though. I'd normally expect the kernel to be either the reader or the writer of the ancillary data, but the current API doesn't really make that distinction.

I'm wondering if it makes sense to split `SocketAncillary` into two distinct types: one for building control messages to pass to the kernel, and one for iterating over control messages received from the kernel.

Reply:
I think https://github.com/rust-lang/rust/pull/86432#discussion_r654332021 is closely related, let's discuss this there :)

Reply:
Added a note :)